source/http/HttpConnection.cpp

```diff
@@ -0,0 +1,363 @@
+#include <aws/crt/io/Bootstrap.h>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        namespace Http
+        {
+            /* this exists so we can keep the shared_ptr for the connection around until the connection has actually
+             * shutdown. */
+            struct ConnectionWrapper
+            {
+                explicit ConnectionWrapper(const HttpClient *client) : connection(nullptr), client(client) {}
+                std::shared_ptr<HttpConnection> connection;
+                const HttpClient *client;
+            };
+
+            void HttpClient::s_onClientConnectionSetup(
+                struct aws_http_connection *connection,
+                int errorCode,
+                void *user_data) noexcept
+            {
+                /**
+                 * Allocate an HttpConnection and seat it to `ConnectionWrapper`'s shared_ptr.
+                 */
+                auto *connectionWrapper = reinterpret_cast<ConnectionWrapper *>(user_data);
+                int retError = errorCode;
+                if (!errorCode)
+                {
+                    Allocator *allocator = connectionWrapper->client->m_allocator;
+
+                    auto *toSeat =
+                        reinterpret_cast<HttpConnection *>(aws_mem_acquire(allocator, sizeof(HttpConnection)));
+                    if (toSeat)
+                    {
+                        toSeat = new (toSeat) HttpConnection(connection, allocator);
+                        connectionWrapper->connection =
+                            std::shared_ptr<HttpConnection>(toSeat, [allocator](HttpConnection *connection) {
+                                connection->~HttpConnection();
+                                aws_mem_release(allocator, reinterpret_cast<void *>(connection));
+                            });
+
+                        connectionWrapper->client->onConnectionSetup(connectionWrapper->connection, errorCode);
+                        return;
+                    }
+
+                    retError = aws_last_error();
+                }
+
+                connectionWrapper->client->onConnectionSetup(nullptr, retError);
+                connectionWrapper->~ConnectionWrapper();
+                aws_mem_release(connectionWrapper->client->m_allocator, reinterpret_cast<void *>(connectionWrapper));
+            }
+
+            void HttpClient::s_onClientConnectionShutdown(
+                struct aws_http_connection *connection,
+                int errorCode,
+                void *user_data) noexcept
+            {
+                (void)connection;
+                /* now that we're shutting down, we can release the internal ref count. */
+                auto *connectionWrapper = reinterpret_cast<ConnectionWrapper *>(user_data);
+                connectionWrapper->client->onConnectionShutdown(connectionWrapper->connection, errorCode);
+                connectionWrapper->connection.reset();
+                connectionWrapper->~ConnectionWrapper();
```

Comment:
This pattern is so common in this CRT, should we do:
```cpp
template <typename T>
void Delete(T* t, Allocator* allocator)
{
    t->~T();
    aws_mem_release(allocator, t);
}
```
and then while we're at it
```cpp
template <typename T, typename ...Args>
T* New(Allocator* allocator, Args... args)
{
    T* t = reinterpret_cast<T*>(aws_mem_acquire(allocator, sizeof(T)));
    if (!t) return nullptr;
    return new(t) T(std::forward<Args...>(args...));
}